workspaces/adventure-pack/goodies/typescript/Iterator.from/index.ts

```diff
@@ -0,0 +1,54 @@
+import "../Iterator.prototype.toIterable";
+
+declare global {
+  interface IteratorConstructor {
+    from<T>(
+      object: Iterator<T> | Iterable<T> | { next(): IteratorResult<T> },
+    ): IterableIterator<T>;
+  }
+
+  const Iterator: IteratorConstructor;
+}
+
+(globalThis as Record<string, unknown>).Iterator ??= {};
+
+(globalThis as Record<string, unknown>).Iterator = {
+  from<T>(
+    object: Iterator<T> | Iterable<T> | { next(): IteratorResult<T> },
+  ): IterableIterator<T> {
+    if (typeof (object as Iterable<T>)[Symbol.iterator] === "function") {
+      const iterator = (object as Iterable<T>)[Symbol.iterator]();
+      if (typeof iterator === "object" && Symbol.iterator in iterator) {
+        return iterator as IterableIterator<T>;
+      } else {
+        return {
+          ...iterator,
+          [Symbol.iterator]() {
+            return this;
+          },
+        } as IterableIterator<T>;
+      }
+    }
+
+    if (typeof (object as Iterator<T>).toIterable === "function") {
```

Comment:
Rather than checking for the presence of a `toIterable` method, we can check if `iteratorPrototype` is in the object's prototype chain, using [`.isPrototypeOf`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isPrototypeOf).

After that check it should be fine to do `.toIterable()` on the object, though once again we can be safe and do `iteratorPrototype.toIterable.call(object)` instead.